src/vs/editor/browser/viewParts/minimap/minimapCharRendererFactory.ts

```diff
@@ -0,0 +1,155 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Constants, MinimapCharRenderer } from 'vs/editor/browser/viewParts/minimap/minimapCharRenderer';
+
+/**
+ * Creates character renderers. It takes a 'scale' that determines how large
+ * characters should be drawn. Using this, it draws data into a canvas and
+ * then downsamples the characters as necessary for the current display.
+ * This makes rendering more efficient, rather than drawing a full (tiny)
+ * font, or downsampling in real-time.
+ */
+export class MinimapCharRendererFactory {
+	private static lastCreated?: MinimapCharRenderer;
+
+	/**
+	 * Creates a new character renderer factory with the given scale.
+	 */
+	public static create(scale: number) {
+		// renderers are immutable. By default we'll 'create' a new minimap
+		// character renderer whenever we switch editors, no need to do extra work.
+		if (this.lastCreated && scale === this.lastCreated.scale) {
+			return this.lastCreated;
+		}
+
+		const factory = MinimapCharRendererFactory.createFromSampleData(MinimapCharRendererFactory.createSampleData().data, scale);
+		this.lastCreated = factory;
+		return factory;
+	}
+
+	/**
+	 * Creates the font sample data, writing to a canvas.
+	 */
+	public static createSampleData(): ImageData {
+		const canvas = document.createElement('canvas');
+		const ctx = canvas.getContext('2d')!;
+
+		canvas.style.height = `${Constants.SAMPLED_CHAR_HEIGHT}px`;
+		canvas.height = Constants.SAMPLED_CHAR_HEIGHT;
+		canvas.width = Constants.CHAR_COUNT * Constants.SAMPLED_CHAR_WIDTH;
+		canvas.style.width = Constants.CHAR_COUNT * Constants.SAMPLED_CHAR_WIDTH + 'px';
+
+		ctx.fillStyle = '#ffffff';
+		ctx.font = `bold ${Constants.SAMPLED_CHAR_HEIGHT}px monospace`;
+		ctx.textBaseline = 'middle';
+		for (let chCode = Constants.START_CH_CODE; chCode <= Constants.END_CH_CODE; chCode++) {
```

Comment:
This doesn't handle unicode or wide characters. Kind of feels like this is just leading to re-implementing xterm.js' texture atlas?

Reply:
Yea, this is something @alexandrudima mentioned. Maybe the texture atlas in xterm could be popped into its own module that vscode can provide to and share between the minimap and xterm instances?

Reply:
Related existing issue that a user noticed on their high DPI display (where we used the 2x4px font) https://github.com/microsoft/vscode/issues/64477